internals/overlord/overlord.go

```diff
@@ -133,6 +141,11 @@ func New(pebbleDir string, restartHandler restart.Handler, serviceOutput io.Writ
 	// the shared task runner should be added last!
 	o.stateEng.AddManager(o.runner)
 
+	// Dry start all managers.
+	if err := o.stateEng.DryStart(); err != nil {
```

Comment:
Should this only happen if `dry` is true?

Reply:
Not really; the dry start mechanism, as discussed on KO026, always runs in order to run initialization tasks that do not incur in unwanted side-effects.

Reply:
Indeed, but it's a bit awkward to see this here at the end of Init.

I think this shoud be run when starting, or on a specific `DryStart` method that only does the dry-starting. This way we also don't need to provide further arguments into the constructor here, and no bools either.

Reply:
The issue of running this when starting (e.g. at the top of `overlord.Loop()`) is that `Loop()` might be called too late and other components (e.g. the daemon) might have already done some state-mutating operations.

The issue of requiring callers to call an `overlord.DryStart()` method manually is that it's easy to forget, and thus misuse the API. Also, `DryStart()` is not an operation you do on the overlord, but rather on the each of the managers. Holding the caller responsible for initializing the managers doesn't seem right to me: the overlord should be the one doing this very internal operation on the managers.

I think running this at the bottom of `overlord.New()` is the sweet spot for these reasons. Also, we don't need any kind of arguments (that argument from previous commits was clearly a bug and a misunderstanding of the problem on my side), because `DryStart()` is meant to be called always.